File: src/main/scala/cn/academy/medicine/MatExtraction.scala
package cn.academy.medicine

import cn.academy.medicine.Properties.Property
import net.minecraft.init.{Blocks, Items}
import net.minecraft.item.{Item, ItemStack}
import Properties._
import net.minecraft.block.Block


object MatExtraction {

  case class ItemMeta(item: Item, meta: Int) {
    lazy val id = {
//      val itemName = Item.itemRegistry.getNameForObject(item)
//      if (item.getHasSubtypes) itemName + "_" + meta else itemName
      Item.getIdFromItem(Item)
    }
  }

  type Recipe = (List[ItemMeta], List[Property])

  def ofBlock(block: Block) = ofItem(Item.getItemFromBlock(block))
  def ofBlock(block: Block, damages: Int*) = ofItem(Item.getItemFromBlock(block), damages: _*)

  def ofItem(item: Item): List[ItemMeta] = ofItem(item, 0 to item.getMaxDamage: _*)
  def ofItem(item: Item, damages: Int*): List[ItemMeta] = damages.map(ItemMeta(item, _)).toList

  val allRecipes: List[Recipe] = List(
    ofItem(Items.APPLE) -> List(Targ_Life, Str_Normal),
    ofItem(Items.POTATO) -> List(Targ_CP),
    ofItem(Items.POISONOUS_POTATO) -> List(Var_Fluct),
    ofItem(Items.WHEAT) -> List(Apply_Continuous_Incr),
    ofItem(Items.REEDS) -> List(Targ_MoveSpeed),
    ofBlock(Blocks.CACTUS) -> List(Str_Mild),
    ofItem(Items.WHEAT_SEEDS) -> List(Apply_Continuous_Incr, Str_Mild),
    ofItem(Items.CARROT) -> List(Str_Weak),
    ofItem(Items.MELON) -> List(Targ_Life),
    ofBlock(Blocks.COCOA) -> List(Targ_Overload, Apply_Continuous_Decr),
    ofItem(Items.EGG) -> List(Apply_Instant_Incr),
    ofItem(Items.MILK_BUCKET) -> List(Apply_Continuous_Incr, Var_Stabilize),
    ofItem(Items.FEATHER) -> List(Targ_Jump),
    ofItem(Items.SPIDER_EYE) -> List(Targ_Cooldown),
    ofItem(Items.BONE) -> List(Str_Mild),
    ofItem(Items.ROTTEN_FLESH) -> List(Var_Fluct),
    ofItem(Items.PORKCHOP) -> List(Targ_CP),
    ofItem(Items.COOKED_PORKCHOP) -> List(Targ_CP),
    ofItem(Items.CHICKEN) -> List(Targ_CP),
    ofItem(Items.COOKED_CHICKEN) -> List(Targ_CP),
    ofItem(Items.BEEF) -> List(Targ_CP),
    ofItem(Items.COOKED_BEEF) -> List(Targ_CP),
    ofItem(Items.FISH) -> List(Apply_Continuous_Incr), // Var_Stablize
    ofItem(Items.COOKED_FISH) -> List(Apply_Continuous_Incr),
    ofItem(Items.ENDER_PEARL) -> List(Targ_CP, Targ_Overload, Var_Fluct),
    ofItem(Items.BLAZE_POWDER) -> List(Targ_Attack, Str_Strong),
    // ofItem(Items.redstone) -> List(Var_Neturalize),
    ofItem(Items.GLOWSTONE_DUST) -> List(Str_Strong, Var_Desens),
    ofItem(Items.NETHER_STAR) -> List(Var_Infinity)
  )

  val allCombinations = allRecipes.flatMap { case (metas, props) =>
      for {
        meta <- metas
        prop <- props
      } yield (meta, prop)
  }

}

File: src/main/java/cn/academy/medicine/MatExtraction.java
package cn.academy.medicine;

import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.Item;

import java.util.*;
import java.util.function.Function;

public class MatExtraction {

  public static class ItemMeta{
      public Item item;
      public int meta;
      public String id;
      public ItemMeta(Item item, int meta) {
          this.item=item;
          this.meta=meta;
          String itemName = Item.itemRegistry.getNameForObject(item);
          id = item.getHasSubtypes()?(itemName + "_" + meta):itemName;
      }

      @Override
      public int hashCode() {
          return id.hashCode();
      }

      @Override
      public boolean equals(Object obj) {
          return obj instanceof ItemMeta && ((ItemMeta) obj).meta==this.meta && ((ItemMeta) obj).item == this.item;
      }
  }

    //type Recipe = (List[ItemMeta], List[Property])

    public static List<ItemMeta> ofBlock(Block block)
    {
        return ofItem(Item.getItemFromBlock(block));
    }

    public static List<ItemMeta> ofBlock(Block block, int... damages){
      return ofItem(Item.getItemFromBlock(block), damages);
    }

    public static List<ItemMeta> ofItem(Item item){
        int[] dmgs=new int[item.getMaxDamage()+1];
        for(int i=0;i<=item.getMaxDamage();i++){
            dmgs[i]=i;
        }
      return ofItem(item,dmgs);
    }
    public static List<ItemMeta> ofItem(Item item, int... damage){
        List<ItemMeta> ret=new ArrayList<>();
        for(int i:damage){
            ret.add(new ItemMeta(item,i));
        }
        return ret;
    }
    static class Recipe{
        public List<ItemMeta> metas;
        public List<Properties.Property> props;
        public Recipe(List<ItemMeta> metas, List<Properties.Property> props){
            this.metas=metas;
            this.props=props;
        }
    }
    public static Map<ItemMeta, List<Properties.Property>> allRecipes = new LinkedHashMap<ItemMeta, List<Properties.Property>>(){{
        for(ItemMeta meta:ofItem(Items.apple))
            put(meta, Arrays.asList(Properties.instance.Targ_Life, Properties.instance.Str_Normal));
        for(ItemMeta meta:ofItem(Items.potato))
            put(meta, Collections.singletonList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.poisonous_potato))
                put(meta, Arrays.asList(Properties.instance.Var_Fluct));
        for(ItemMeta meta:ofItem(Items.wheat))
                put(meta, Arrays.asList(Properties.instance.Apply_Continuous_Incr));
        for(ItemMeta meta:ofItem(Items.reeds))
                put(meta, Arrays.asList(Properties.instance.Targ_MoveSpeed));
        for(ItemMeta meta:ofBlock(Blocks.cactus))
                put(meta, Arrays.asList(Properties.instance.Str_Mild));
        for(ItemMeta meta:ofItem(Items.wheat_seeds))
                put(meta, Arrays.asList(Properties.instance.Apply_Continuous_Incr, Properties.instance.Str_Mild));
        for(ItemMeta meta:ofItem(Items.carrot))
                put(meta, Arrays.asList(Properties.instance.Str_Weak));
        for(ItemMeta meta:ofItem(Items.melon))
                put(meta, Arrays.asList(Properties.instance.Targ_Life));
        for(ItemMeta meta:ofBlock(Blocks.cocoa))
                put(meta, Arrays.asList(Properties.instance.Targ_Overload, Properties.instance.Apply_Continuous_Decr));
        for(ItemMeta meta:ofItem(Items.egg))
                put(meta, Arrays.asList(Properties.instance.Apply_Instant_Incr));
        for(ItemMeta meta:ofItem(Items.milk_bucket))
                put(meta, Arrays.asList(Properties.instance.Apply_Continuous_Incr, Properties.instance.Var_Stabilize));
        for(ItemMeta meta:ofItem(Items.feather))
                put(meta, Arrays.asList(Properties.instance.Targ_Jump));
        for(ItemMeta meta:ofItem(Items.spider_eye))
                put(meta, Arrays.asList(Properties.instance.Targ_Cooldown));
        for(ItemMeta meta:ofItem(Items.bone))
                put(meta, Arrays.asList(Properties.instance.Str_Mild));
        for(ItemMeta meta:ofItem(Items.rotten_flesh))
                put(meta, Arrays.asList(Properties.instance.Var_Fluct));
        for(ItemMeta meta:ofItem(Items.porkchop))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.cooked_porkchop))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.chicken))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.cooked_chicken))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.beef))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.cooked_beef))
                put(meta, Arrays.asList(Properties.instance.Targ_CP));
        for(ItemMeta meta:ofItem(Items.fish))
                put(meta, Arrays.asList(Properties.instance.Apply_Continuous_Incr)); // Var_Stablize
        for(ItemMeta meta:ofItem(Items.cooked_fished))
                put(meta, Arrays.asList(Properties.instance.Apply_Continuous_Incr));
        for(ItemMeta meta:ofItem(Items.ender_pearl))
                put(meta, Arrays.asList(Properties.instance.Targ_CP, Properties.instance.Targ_Overload,
                        Properties.instance.Var_Fluct));
        for(ItemMeta meta:ofItem(Items.blaze_powder))
                put(meta, Arrays.asList(Properties.instance.Targ_Attack, Properties.instance.Str_Strong));
        // for(ItemMeta meta:ofItem(Items.redstone))
        //        put(meta, Arrays.asList(Var_Neturalize));
        for(ItemMeta meta:ofItem(Items.glowstone_dust))
                put(meta, Arrays.asList(Properties.instance.Str_Strong, Properties.instance.Var_Desens));
        for(ItemMeta meta:ofItem(Items.nether_star))
                put(meta, Arrays.asList(Properties.instance.Var_Infinity));
    }};

    public static boolean checkIfReciped(Item item, int i){
        return allRecipes.containsKey(new ItemMeta(item, i));
    }

}
